File: solutions/go/06/inclusion_exclusion/main.go
package main

import "fmt"

func countCoprime(n int) int {
	primes := []int{2, 3}
	ans := n
	m := len(primes)
	for mask := 1; mask < 1<<m; mask++ {
		bits := 0
		prod := 1
		for i := 0; i < m; i++ {
			if mask>>i&1 == 1 {
				bits++
				prod *= primes[i]
			}
		}
		if bits%2 == 1 {
			ans -= n / prod
		} else {
			ans += n / prod
		}
	}
	return ans
}

func main() {
	fmt.Println(countCoprime(30))
}

File: solutions/python/06/inclusion_exclusion/main.py
def count_coprime(n: int) -> int:
    """1..n のうち 2 でも 3 でも割り切れない個数（包除原理）"""
    primes = [2, 3]
    ans = n
    m = len(primes)
    for mask in range(1, 1 << m):
        bits = bin(mask).count("1")
        prod = 1
        for i in range(m):
            if mask >> i & 1:
                prod *= primes[i]
        if bits % 2:
            ans -= n // prod
        else:
            ans += n // prod
    return ans


def main() -> None:
    print(count_coprime(30))


if __name__ == "__main__":
    main()
